src/torchjd/sparse/_linalg.py

```diff
@@ -0,0 +1,216 @@
+            U[:, 0:col] -= factors.unsqueeze(0) * U[:, col].unsqueeze(1)
+            V[col, :] += factors @ V[0:col, :]
+
+        col += 1
+
+    # TODO: Should actually make 2 functions, one for full and one for reduced
+    if reduced:
+        rank = _get_hermite_factor_rank(H)
+
+        H = H[:, :rank]
+        U = U[:, :rank]
+        V = V[:rank, :]
+
+    return H, U, V
+
+
+def compute_gcd(S1: Tensor, S2: Tensor) -> tuple[Tensor, Tensor, Tensor]:
+    """
+    Computes a GCD and the projection factors, i.e.
+    S1 = G @ K1
+    S2 = G @ K2
+    with G having minimal rank r.
+
+    Args:
+        S1, S2: torch.Tensors (m x n1), (m x n2)
+
+    Returns:
+        G: (m x r) The Matrix GCD (Canonical Base)
+        K1: (r x n1) Factors for S1
+        K2: (r x n2) Factors for S2
+
+    Implementation logic:
+    The concatenated matrix [S1 | S2] spans exactly the sum of the lattices generated by S1 and S2.
+    This is because S1 @ u1 + S2 @ u2 = [S1 | S2] @ [u1.T | u2.T].T
+    The reduced HNF decomposition of [S1 | S2] yields G, U, V where the G.shape[1] is the rank of
+    [S1 | S2] and [S1 | S2] = G @ V. This means that
+    S1 = G @ V[:, :m1]
+    S2 = G @ V[:, m1:]
+    This is the target common factorization. It is the greatest as the lattice spanned by G is the
+    same as that spanned by [S1 | S2].
+    """
+    assert S1.shape[0] == S2.shape[0]
+    m, n1 = S1.shape
+
+    A = torch.cat([S1, S2], dim=1)
+    G, _, V = hnf_decomposition(A, True)
+
+    K1 = V[:, :n1]
+    K2 = V[:, n1:]
+
+    return G, K1, K2
+
+
+def compute_lcm(S1: Tensor, S2: Tensor) -> tuple[Tensor, Tensor, Tensor]:
+    """
+    Computes a LCM and the projection multipliers, i.e.
+    L = S1 @ M1 = S2 @ M2
+    with L having maximal rank r.
+
+    Args:
+        S1, S2: torch.Tensors (m x n1), (m x n2)
+
+    Returns:
+        L:  (m x r)  The Matrix LCM
+        M1: (n1 x r) Multiplier for S1
+        M2: (n2 x r) Multiplier for S2
+
+    Implementation logic:
+    The lattice kernel of the concatenated matrix [S1 | -S2] is the set of all vectors
+    [u1.T | u2.T].T such that S1 @ u1 - S2 @ u2 = 0, or equivalently S1 @ u1 = S2 @ u2.
+    This means that the image of the components of the kernel through S1 and S2 respectively are the
+    same which is exactly the intersection of the lattices generated by S1 and S2.
+    The full HNF decomposition of [S1 | -S2] yields H, U, V such that
+    H = [S1 | -S2] @ U
+    If [S1 | -S2] has rank r', then every column of H after the first r' contain only zeros, and
+    therefore U[:, r':] spans the kernel of [S1 | -S2]. We have
+    S1 @ U[:n1, r':] = S2 @ U[n1:, r':]
+    which yields the desired decomposition with r=n1+n2-r'.
+    """
+    assert S1.shape[0] == S2.shape[0]
+    m, n1 = S1.shape
+
+    B = torch.cat([S1, -S2], dim=1)
+    H, U, _ = hnf_decomposition(B, False)
+
+    rank = _get_hermite_factor_rank(H)
+    M2 = U[n1:, -rank:]
+    M1 = U[:n1, -rank:]
```

Comment:
@ValerianRey if you don't mind, I prefer like that, LGTM otherwise (good catch, thx).
```suggestion
    M1 = U[:n1, -rank:]
    M2 = U[n1:, -rank:]
```